test/unit/org/apache/cassandra/utils/memory/NativeAllocatorTest.java

```diff
@@ -23,105 +23,140 @@
-                    // confirm adjustment works
-                    allocator.offHeap().adjust(-10, group);
-                    Assert.assertEquals(0, allocator.offHeap().owns());
-                    allocator.offHeap().adjust(10, group);
-                    Assert.assertEquals(10, allocator.offHeap().owns());
-                    // confirm we cannot allocate negative
-                    boolean success = false;
-                    try
-                    {
-                        allocator.offHeap().allocate(-10, group);
-                    }
-                    catch (AssertionError e)
-                    {
-                        success = true;
-                    }
-                    Assert.assertTrue(success);
-                    Uninterruptibles.sleepUninterruptibly(10L, TimeUnit.MILLISECONDS);
-                    Assert.assertEquals(1, isClean.getCount());
-
-                    // allocate above watermark, check cleaned
-                    allocator.allocate(70, group);
-                    Assert.assertEquals(80, allocator.offHeap().owns());
-                    canClean.countDown();
-                    try
-                    {
-                        isClean.await(10L, TimeUnit.MILLISECONDS);
-                    }
-                    catch (InterruptedException e)
-                    {
-                        throw new AssertionError();
-                    }
-                    Assert.assertEquals(0, isClean.getCount());
-                    Assert.assertEquals(0, allocator.offHeap().owns());
-
-                    // allocate above limit, check we block until "marked blocking"
-                    exec.schedule(markBlocking, 10L, TimeUnit.MILLISECONDS);
-                    allocator.allocate(110, group);
-                    Assert.assertNotNull(barrier.get());
-                    Assert.assertEquals(110, allocator.offHeap().owns());
-                }
-            };
-            exec.submit(run).get();
-        }
-    }
+                isClean.await(10L, TimeUnit.MILLISECONDS);
+            }
+            catch (InterruptedException e)
+            {
+                throw new AssertionError();
+            }
+            Assert.assertEquals(0, isClean.getCount());
+            verifyUsedReclaiming(0, 0);
 
+            // allocate, then set discarding, then allocated some more
+            allocator.allocate(30, group);
+            verifyUsedReclaiming(30, 0);
+            allocator.setDiscarding();
+            Assert.assertFalse(allocator.isLive());
+            verifyUsedReclaiming(30, 30);
+            allocator.allocate(50, group);
+            verifyUsedReclaiming(80, 80);
+
+            // allocate above limit, check we block until "marked blocking"
+            exec.schedule(markBlocking, 10L, TimeUnit.MILLISECONDS);
+            allocator.allocate(30, group);
+            Assert.assertNotNull(barrier.get());
+            verifyUsedReclaiming(110, 110);
```

Comment:
We're testing a lot of things in `testBookKeeping()`. I'd suggest breaking some of them out into their own (more specifically named) test methods that can make a more focused set of assertions. You've already pulled out some initialization into `@Before`, which I imagine will make it easier...